op-acceptance-tests/tests/base/reth_geth_divergence_test.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+)
+
+// TestRethGethDivergence checks that reth and geth do not diverge in terms of state root.
+// geth is considered the canonical/correct implementation.
+func TestRethGethDivergence(t *testing.T) {
+	systest.SystemTest(t, rethGethDivergenceTestScenario())
+}
+
+func rethGethDivergenceTestScenario() systest.SystemTestFunc {
+	const (
+		// testTimeout is the maximum time allowed for the test to run
+		testTimeout = 60 * time.Second
+
+		// blockCheckCount is the number of recent blocks to check for divergence
+		blockCheckCount = 5
+	)
+
+	return func(t systest.T, sys system.System) {
+		logger := testlog.Logger(t, log.LevelInfo)
+		logger.Info("Started reth-geth divergence test")
+
+		// Check all L2 chains
+		for i, chain := range sys.L2s() {
+			chainIndex := i
+			currentChain := chain
+			t.Run(fmt.Sprintf("Chain_%d", chainIndex), func(t systest.T) {
+				t.Parallel()
+				checkChainForDivergence(t, currentChain, logger.New("chain", chainIndex), testTimeout, blockCheckCount)
+			})
+		}
+	}
+}
+
+// checkChainForDivergence checks a single chain for reth/geth divergence
+func checkChainForDivergence(
+	t systest.T,
+	chain system.L2Chain,
+	logger log.Logger,
+	testTimeout time.Duration,
+	blockCheckCount uint64,
+) {
+	// Create context with timeout
+	ctx, cancel := context.WithTimeout(t.Context(), testTimeout)
+	defer cancel()
+
+	// Find geth and reth nodes
+	gethNode, rethNode, err := findGethAndRethNodes(chain, logger)
+	if err != nil {
+		t.Skip(err.Error())
+		return
+	}
+
+	logger.Info("Found both geth and reth nodes",
+		"geth", gethNode.Name(),
+		"reth", rethNode.Name())
+
+	// Get clients for both nodes
+	gethClient, rethClient, err := getNodeClients(gethNode, rethNode)
+	if err != nil {
+		t.Fatalf("Failed to get clients: %v", err)
+		return
+	}
+
+	// Determine the block range to check
+	startBlock, endBlock, err := getBlockRangeToCheck(ctx, gethClient, blockCheckCount)
+	if err != nil {
+		t.Fatalf("Failed to determine block range: %v", err)
+		return
+	}
+
+	// Check each block for divergence
+	for blockNumber := startBlock; blockNumber <= endBlock; blockNumber++ {
+		checkBlockForDivergence(ctx, t, gethClient, rethClient, blockNumber, logger)
+	}
+
+	logger.Info("State root divergence check passed for all blocks",
+		"start_block", startBlock,
+		"end_block", endBlock)
+}
+
+// findGethAndRethNodes finds the geth and reth nodes in the chain
+func findGethAndRethNodes(chain system.L2Chain, logger log.Logger) (gethNode, rethNode system.Node, err error) {
+	for _, node := range chain.Nodes() {
+		nodeName := node.Name()
+		logger.Debug("Found node", "name", nodeName)
+
+		if isGethNode(nodeName) {
+			gethNode = node
+			logger.Debug("Identified geth node", "name", nodeName)
+		} else if isRethNode(nodeName) {
```

Comment:
This function seems to find the last known reth nodes in the chain.

IMO we should return a slice of all the reth nodes because we would want to check divergence for all of them.

Reply:
Agreed. Done.